### ISSUE TITLE
forward_socks fail when visit speedtest.net

### ISSUE THREAD

**Issue by pttdev** [ronf/asyncssh#159]
I create a socks5 with this simple code:

```
import pprint
import sys
import asyncio
import asyncssh

async def ssh_connect():
    conn, client = await asyncio.wait_for(asyncssh.create_connection(
        None, '123.123.123.123',
        known_hosts=None,
        username='root',
        password='1234'),
        10
    )
    async with conn:
        listener = await asyncio.wait_for(conn.forward_socks('127.0.0.1', 8080), 600)
        return await listener.wait_closed()

try:
    asyncio.get_event_loop().run_until_complete(ssh_connect())
except (OSError, asyncssh.Error) as exc:
    print(str(exc))
```
Next i visit normal website and it work just fine. But when i visit speedtest.net, the socks stop working without any errors. Does anyone ran into this trouble and know how to fix it? please tell me know. Thanks you very much!

**Comment by ronf** [comment]
Thanks for the report. I think I've been able to reproduce this here. It appears to be a race condition related to support for half-closed TCP connections. When the other side has fully closed and we try to half-close (before the application has said to fully close), a socket error is returned. When I increased the debugging level, I saw the following in this case:

```
DEBUG:asyncssh:[conn=0] Uncaught exception
Traceback (most recent call last):
  File "/Users/ronf/src/asyncssh/tmp/asyncssh/connection.py", line 616, in data_received
    while self._inpbuf and self._recv_handler():
  File "/Users/ronf/src/asyncssh/tmp/asyncssh/connection.py", line 836, in _recv_packet
    processed = handler.process_packet(pkttype, seq, packet)
  File "/Users/ronf/src/asyncssh/tmp/asyncssh/packet.py", line 207, in process_packet
    self._packet_handlers[pkttype](self, pkttype, pktid, packet)
  File "/Users/ronf/src/asyncssh/tmp/asyncssh/channel.py", line 554, in _process_eof
    self._flush_recv_buf()
  File "/Users/ronf/src/asyncssh/tmp/asyncssh/channel.py", line 291, in _flush_recv_buf
    if (not self._session.eof_received() and
  File "/Users/ronf/src/asyncssh/tmp/asyncssh/forward.py", line 98, in eof_received
    self._peer.write_eof()
  File "/Users/ronf/src/asyncssh/tmp/asyncssh/forward.py", line 46, in write_eof
    self._transport.write_eof()
  File "/Users/ronf/src/asyncssh/tmp/asyncio/selector_events.py", line 754, in write_eof
    self._sock.shutdown(socket.SHUT_WR)
OSError: [Errno 57] Socket is not connected
```

You can try the following patch - it seems to work here for me:

```patch
index 4c7cda5..8d6d641 100644
--- a/asyncssh/forward.py
+++ b/asyncssh/forward.py
@@ -43,7 +43,10 @@ class SSHForwarder:
     def write_eof(self):
         """Write end of file to the transport"""
 
-        self._transport.write_eof()
+        try:
+            self._transport.write_eof()
+        except OSError:
+            pass
 
     def was_eof_received(self):
         """Return whether end of file has been received or not"""
```

I'll get this fix into the next release.

**Comment by pttdev** [comment]
Thank you very much for quick fix. Confirm that it work, hope you publish new release soon.
Have a good day!

**Comment by ronf** [comment]
This fix is now available in the "develop" branch if you don't want to wait for the next official release, but I'll try to get that out soon.

**Comment by ronf** [comment]
This is now available in AsyncSSH 1.14.0.
